autopilot/scanner/scanner.go

```diff
@@ -216,18 +179,65 @@ func (s *scanner) scanHosts(ctx context.Context, w WorkerRHPScan, hosts chan sca
 		}
 	}
 
-	// launch all workers
-	var wg sync.WaitGroup
-	for t := 0; t < s.scanThreads; t++ {
-		wg.Add(1)
-		go func() {
-			worker()
-			wg.Done()
-		}()
-	}
+	var exhausted bool
+	for !exhausted {
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.shutdownChan:
+			return
+		default:
+		}
+
+		// synchronisation vars
+		jobs := make(chan scanJob)
+		workersDone := make(chan struct{})
+		joinWorkers := func() {
+			close(jobs)
+			<-workersDone
+		}
+
+		// launch all workers for this batch
+		activeWorkers := int32(s.scanThreads)
+		for i := 0; i < s.scanThreads; i++ {
+			go func() {
+				worker(jobs)
+				if atomic.AddInt32(&activeWorkers, -1) == 0 {
+					close(workersDone) // close when all workers are done
+				}
+			}()
+		}
 
-	// wait until they're done
-	wg.Wait()
+		// fetch batch
+		hosts, err := s.hs.Hosts(ctx, api.HostOptions{
+			MaxLastScan: api.TimeRFC3339(cutoff),
+			Offset:      0,
+			Limit:       s.scanBatchSize,
+		})
+		if err != nil {
+			s.logger.Errorf("could not get hosts for scanning, err: %v", err)
+			joinWorkers()
+			break
+		}
+		exhausted = len(hosts) < s.scanBatchSize
+
+		// send batch to workers
+		for _, h := range hosts {
+			select {
+			case jobs <- scanJob{
+				hostKey: h.PublicKey,
+				hostIP:  h.NetAddress,
+			}:
+			case <-ctx.Done():
+				continue
+			case <-s.shutdownChan:
+				continue
+			case <-workersDone:
+				continue
+			}
+		}
+		joinWorkers()
+	}
 
 	s.statsHostPingMS.Recompute()
```

Comment:
`nit` doesn't matter much but we could `defer` this since we swapped `break` statements for `return`